heron/metricscachemgr/src/java/com/twitter/heron/metricscachemgr/metricscache/CacheCore.java

```diff
@@ -0,0 +1,543 @@
+        } // end for: component
+      } // end for: metric
+      return response;
+    }
+  }
+
+  private void getRawMetrics(List<MetricsCacheQueryUtils.MetricTimeRangeValue> metricValue,
+                             long startTime, long endTime, long bucketId,
+                             MetricsFilter.MetricAggregationType type) {
+    LOG.info("getRawMetrics " + startTime + " " + endTime);
+    Long startKey = cacheMetric.floorKey(startTime);
+    for (Long key = startKey != null ? startKey : cacheMetric.firstKey();
+         key != null && key <= endTime;
+         key = cacheMetric.higherKey(key)) {
+      LinkedList<MetricDatapoint> bucket = cacheMetric.get(key).get(bucketId);
+
+      if (bucket != null) {
+        for (MetricDatapoint datapoint : bucket) {
+          if (startTime <= datapoint.timestamp && datapoint.timestamp <= endTime) {
+            // per data point
+            MetricsCacheQueryUtils.MetricTimeRangeValue rangeValue =
+                new MetricsCacheQueryUtils.MetricTimeRangeValue();
+            rangeValue.startTime = datapoint.timestamp;
+            rangeValue.endTime = datapoint.timestamp;
+            rangeValue.value = datapoint.value;
+            metricValue.add(rangeValue);
+          }
+        } // end bucket
+      }
+
+    } // end tree
+  }
+
+  // we assume the metric value is Double: compatible with tmaster
+  @SuppressWarnings("fallthrough")
+  private void getMinuteMetrics(List<MetricsCacheQueryUtils.MetricTimeRangeValue> metricValue,
+                                long startTime, long endTime, long bucketId,
+                                MetricsFilter.MetricAggregationType type) {
+    LOG.info("getMinuteMetrics " + startTime + " " + endTime);
+    Long startKey = cacheMetric.floorKey(startTime);
+    for (Long key = startKey != null ? startKey : cacheMetric.firstKey();
+         key != null && key <= endTime;
+         key = cacheMetric.higherKey(key)) {
+      LinkedList<MetricDatapoint> bucket = cacheMetric.get(key).get(bucketId);
+
+      if (bucket != null) {
+        // per bucket
+        long countAvg = 0;
+
+        MetricsCacheQueryUtils.MetricTimeRangeValue rangeValue =
+            new MetricsCacheQueryUtils.MetricTimeRangeValue();
+        rangeValue.startTime = Long.MAX_VALUE;
+        rangeValue.endTime = 0;
+
+        double result = 0;
+        for (MetricDatapoint datapoint : bucket) {
+          if (startTime <= datapoint.timestamp && datapoint.timestamp <= endTime) {
+            switch (type) {
+              case AVG:
+                countAvg++;
+              case SUM:
+                result += Double.parseDouble(datapoint.value);
+                break;
+              case LAST:
+                if (rangeValue.endTime < datapoint.timestamp) {
+                  rangeValue.value = datapoint.value;
+                }
+                break;
+              case UNKNOWN:
+              default:
+                LOG.warning("Unknown metric type " + type);
```

Comment:
Can you provide more context to the log reader re what is being attempted with the unknown metric type? Seeing 'Unknown metric type foo' in the logs doesn't convey much. Same for logging in getAggregatedMetrics method.